### ISSUE TITLE
[Feature] Cameo/NonSex tag or list for performers in scene

### ISSUE THREAD

**Issue by Emilo2** [stashapp/stash-box#342]
**Is your feature request related to a problem? Please describe.**
Many scenes have extra performers listed that don't participate in the main act.

**Describe the solution you'd like**
IAFD has NonSex tag added to performer that is not participating in the main act.

**Describe alternatives you've considered**
Separate list for cameo/nonsex roles would also work, like the Director is currently.

**Additional context**
We want to have as much data as possible for scenes. Leaving listed or credited performers out because they don't actually perform anything is not very nice.

**Comment by DogmaDragon** [comment]
This was brought up on Discord yesterday and someone had a good point. There is no need for a different role system since we already have a scene alias that can easily be used for this purpose. 

E.g. 
![Screenshot_164](https://user-images.githubusercontent.com/103123951/166254056-d17e5b65-248b-43ce-b703-c0551bf00426.png)

All we need is a guideline to formalize its usage.

**Comment by stg-annon** [comment]
I think this would fit better as an attribute under `PerformerAppearance` this way they could be filtered as desired and they would have a defined set of roles that could be added to later.

for example if stash users don't care about non-sex roles they could not request them from stash-box if they were incorporated into the alias it would always return them for scenes with many background performers and few main performers this would be an issue

**Comment by AdultSun** [comment]
This came up again recently on Discord, where we discussed the idea of having multiple "performance flags" to check. IAFD uses a similar system but many of theirs are covered by basic scene tags on StashDB already. Their full list is here: 
https://www.iafd.com/glossary.asp

Both IAFD and IMDB have flags to list a performer as "uncredited" as well. We realized that this would also be a useful performance flag to have, especially with the way our "Update scene performance aliases on merged performers to old performer name" feature works when renaming/merging performers. If we leave the scene alias field blank (which is most often the case with uncredited performers currently) and the performer is renamed/merged with that option checked, the old name will automatically be used as a scene alias implying that this was how they were credited when it actually just reflects StashDB's edit history and nothing else. We could start using `Uncredited` as a scene alias instead like DogmaDragon's suggestion of adding ` (NonSex)`, but those would likely be temporary solutions.

So if we already want to add two different performance flags, then we might as well brainstorm as many possible flags to include in a system like this. I'm already imagining this as a multi-select dropdown system since we could easily have a performer who is both "NonSex" and "Uncredited" at the same time. Here's all the ones I can think of, most of them borrowed from IAFD's system, roughly in order of importance:

- `Uncredited`: Performers who have been identified but aren't officially listed by the studio, often cameos or male performers in straight scenes. This is the only one that I think should have additional functionality. Similar to the "organized" flag in Stash, we'd want "uncredited" performers to be exempt from "update scene alias" actions when merging or renaming the performer.
- `Non-Sex`/`NonSex`/`Cameo`: Not directly involved in the action, would cover cameos during the intro/outro of a scene or even cuckold/voyeur/exhibitionist setups.
- `Non-Nude`/`NonNude` vs. `Nude`: Could be helpful to clarify whether a non-sex performer stays fully clothed or not.
- `Top`: The performer doing the penetrating, helpful for gay/bi/trans/strap-on scenes.
- `Bottom`: The performer being penetrated, helpful for gay/bi/trans/strap-on scenes.
- `Versatile`/`Vers`: Performer both penetrates and is penetrated, helpful for gay/bi/trans/strap-on scenes. The term "Switch" could be used as well, but probably better fits BDSM scenes instead.
- `Dom`/`Dominant`: Performer taking a dominant role, helpful when `MaleDom` and `FemDom` scene tags aren't enough to fully describe each performer's role. "Top" could be used here as well, but better fits gay/bi/trans/strap-on scenes.
- `Sub`/`Submissive`: Performer taking a submissive role, helpful when `MaleDom` and `FemDom` scene tags aren't enough to fully describe each performer's role. "Bottom" could be used here as well, but better fits gay/bi/trans/strap-on scenes.
- `Switch`: Could be used for performers that play both a dominant and submissive role within the same scene, not sure how often this happens. "Versatile" could be used here as well, but better fits gay/bi/trans/strap-on scenes.
- `Blowjob Only`/`BJ Only`/`BJOnly`: We already have a `Blowjob Only` scene tag so this would only be relevant when one performer is BJOnly but another is not.
- `Lesbian Only`/`Lez Only`/`LezOnly`: Female performers who only interact with other female performers in a scene, only relevant when used for mixed-gender scenes since we can just use our regular lesbian tags for female-only scenes.
- `Masturbation Only`/`MastOnly`: Only sexual contact is from themselves, only relevant for non-solo scenes.
- `Body Doubled`: For performers who are swapped out with a stand-in for a particular shot/action that's hidden by the editing, really only relevant for older movies.
- `Body Doubling`: For the performers who are used as the stand-in for another performer, not always known and would almost definitely be `Uncredited` as well, and again only relevant for older movies.
- `Clip`/`Archival`: Performer only appears in previously released "archival" footage, mostly relevant for older movies when recycling scenes was common. Occasionally hardcore footage is included as a flashback/recap to previous scenes in a series in modern releases as well. Could also be used for modern examples of remasters/redistributions/re-releases but we have scene tags for that already.
- `Peeing`/`Pee`: Performer shown urinating, only helpful in non-solo scenes when other performers are not peeing.
- `Peed On`/`Golden Shower`/`GS`: Performer shown being urinated on, only helpful when it doesn't apply to other performers in a scene.

IAFD has a lot of flags that we already have a scene tag for and are only relevant if it applies to one performer but not another in group scenes. I don't believe any of these are necessary to add to StashDB but I'll list them here anyway:
- `Anal`
- `Creampie`
- `DP`
- `DPP`
- `DAP`
- `Facial`
- `Fisting`
- `Squirt`
- `IR`
- `a2m`/`Ass to Mouth
- `Swallow`
- `CumSwap`/`Cum Swap`
- `Bald`: "Appears with no pubic hair. Different from shaved."
- `Shaved`: "We see the pubic hair come off."
- `VR`: Always applies to the entire scene so not helpful as a performance flag at all.
- `DVD Only`/`DVDOnly`: Only included in a physical media release and not digital releases, sounds more like a movie tag than a performer flag.

**Comment by echo6ix** [comment]
> IAFD has a lot of flags that we already have a scene tag for and are only relevant if it applies to one performer but not another in group scenes. I don't believe any of these are necessary to add to StashDB but I'll list them here anyway:
> 
> * `Anal`
> * `Creampie`
> [...]

Yes. +1 not necessary. IMO this will create excessive redundancy and confusion with the existing tags. Please do not add any of those that can already be covered by the scene tags.

**Comment by echo6ix** [comment]
Concept of how this might work with the UI

![image](https://user-images.githubusercontent.com/37937507/197658583-bbddb94f-3f4e-47a8-9de9-af253793ecad.png)

**Comment by echo6ix** [comment]
From @AdultSun on [Discord](https://discord.com/channels/559159668438728723/642050893549928449/1034271449545977856),
> Another part of the UI I'm not sure about is how those tags/flags should be displayed in the scene details page. If it's a multi-select list, it could get pretty crowded very quickly alongside scene aliases and disambiguations. We may want to hide some of those on hover or something

I figure something like this could work on the scene page under the video thumbnail

![image](https://user-images.githubusercontent.com/37937507/197660936-fa3f1149-8f57-46bc-8fec-28f652e9e4d1.png)

The current format is `Alias Name (Primary Name)`, and with the inclusion of flags, presumably in brackets, the amount of brackets being displayed on this line would get to bit overwhelming, especially if there's a disambiguation. Just changing the format to something like this might make more sense:

- Primary Name (Disambiguation in parenthesis) as **Alias Name** _(flags italicized in parenthesis)_

Edit: Another more minimalist display option on the scenes details page could be to omit the info icon for hyperlinked ellipses when there is greater than 2 or 3 flags with a popover on hover like so...

![image](https://user-images.githubusercontent.com/37937507/197869765-7e15ee05-bee9-4ea8-a511-bbdbed9248c5.png)

**Comment by DogmaDragon** [comment]
> * `Uncredited`: Performers who have been identified but aren't officially listed by the studio, often cameos or male performers in straight scenes. This is the only one that I think should have additional functionality. Similar to the "organized" flag in Stash, we'd want "uncredited" performers to be exempt from "update scene alias" actions when merging or renaming the performer.

I don't see any reason for this flag to be visible on the front end. Make it togglable on the edit page and invisible on the scene details page. 
> * `Non-Sex`/`NonSex`/`Cameo`: Not directly involved in the action, would cover cameos during the intro/outro of a scene or even cuckold/voyeur/exhibitionist setups.

It needs a strict definition to be of any use. By StashDB definition there are plenty of sex acts <https://stashdb.org/categories/feca7511-ac91-42c0-a032-8fb8f34350bc> that other sites with `non-sex` flags would exclude.

---
I think flags should be unique and have a non-trivial purpose. Most of the proposed flags already provide the same information via the tags system. And unlike tags, these flags would have to be toggled manually so if people have to go through a list of 20 flags each time they submit a new scene it will be ignored and misused.

**Comment by AdultSun** [comment]
I agree with all those statements.

I'd say the first 6 flags in my list are the most important that I'd like to see included. There's an argument for everything else in that first group, but the rest of those are of limited relevance as I tried to explain for each one already. I don't think anything in the second group (the same one Echo mentioned) is useful enough to be worth including.

I also mentioned in Discord that I would lean towards hiding all the new "flags" behind that [i] icon in Echo's diagram. If any are visible by default, I think the most important one would be our "Non-Sex" flag but personally I'd rather treat all of them the same and hide that too.

Speaking of, to me that flag is most useful for indicating "this performer appears in the scene, but isn't really *involved* in the scene" if that makes sense. IAFD defines "NonSex" on their glossary page as "Has no sex; may or may not be nude," which doesn't seem particularly helpful to me. How do you define sex, how does it apply to BDSM "handler" scenes, how does it apply to softcore "pretty girl" posing scenes, solo masturbation scenes, lesbian scenes, etc. For what I'm thinking, it may be more accurate to use the term `Cameo` instead of `Non-Sex` so it's not confused with the [No Sex](https://stashdb.org/tags/cd552afe-c6ba-41d8-8fa3-0a77d2ab87f4) scene tag we already have. Meaning, it's intended for performers who are visible or present in the scene but not actively involved in the main "action" in any way. The content covered by StashDB is just too varied to easily define "action" any more clearly than that, I don't see any way around it. So if it's a BDSM scene, not engaged in the flogging or whatever. If it's a solo nude posing scene, then not solo nude posing themselves I guess. Neither of those examples would have "sex" as defined by IAFD, but it doesn't seem helpful to me to flag every single performer in those types of scenes with a `NonSex` performance flag. It would be more fitting to slap a `No Sex` tag on the scene instead and move on.

**Comment by DogmaDragon** [comment]
> Speaking of, to me that flag is most useful for indicating "this performer appears in the scene, but isn't really _involved_ in the scene" if that makes sense. IAFD defines "NonSex" on their glossary page as "Has no sex; may or may not be nude," which doesn't seem particularly helpful to me. How do you define sex, how does it apply to BDSM "handler" scenes, how does it apply to softcore "pretty girl" posing scenes, solo masturbation scenes, lesbian scenes, etc. For what I'm thinking, it may be more accurate to use the term `Cameo` instead of `Non-Sex` so it's not confused with the [No Sex](https://stashdb.org/tags/cd552afe-c6ba-41d8-8fa3-0a77d2ab87f4) scene tag we already have. Meaning, it's intended for performers who are visible or present in the scene but not actively involved in the main "action" in any way. The content covered by StashDB is just too varied to easily define "action" any more clearly than that, I don't see any way around it. So if it's a BDSM scene, not engaged in the flogging or whatever. If it's a solo nude posing scene, then not solo nude posing themselves I guess. Neither of those examples would have "sex" as defined by IAFD, but it doesn't seem helpful to me to flag every single performer in those types of scenes with a `NonSex` performance flag. It would be more fitting to slap a `No Sex` tag on the scene instead and move on.

I like `Cameo` more than `Non-Sex`. For me, the purpose of this flag would be to indicate the performers that only exist for plot purposes and not the action. E.g. the scenes where step-x gets caught by someone. 

I'm a bit torn about cuckold/voyeur/exhibitionist content. For me the difference is between do they participate in any way or just watch. But it's a bit harder to moderate especially if the person submitting hasn't seen the scene. 

Flogging is a sexual act, the same with scenes where only toys are used. It's rare that rigger doesn't participate in any other way. So I would say the flag should only be used for scenes the rigger only does the rigging and doesn't participate in any other way. 

When it comes to solo content, I think by definition it's sexual so it would be pointless to mark every scene with the flag. 

In short, my threshold is plot purposes only = flag, any sexual act = no flag.

**Comment by echo6ix** [comment]
IMDB has some useful guidelines and ideas regarding this topic https://help.imdb.com/article/contribution/filmography-credits/cast-acting-credits-guidelines/GH3JZC74FVYKKFMD?ref_=helpsrall

Their UI wasn't too different from my concept, except what I labeled a _flag_, they labeled an _attribute_, which I think is preferential.

![image](https://github.com/user-attachments/assets/40a41984-6bd3-480c-9f90-50b9f22d3e8d)

IMDB's list of valid attributes are located [here](https://help.imdb.com/article/contribution/filmography-credits/cast-acting-credits-guidelines/GH3JZC74FVYKKFMD?ref_=helpsrall#attribute) but the utility of these will vary depending on the stash-box use-case.

Which is why a stash-box, specifically the administrators, should have the ability to customize (modify, remove, add) their list of valid attributes available to the users. So the need to think of a complete list attributes upon implementing this feature is low priority. Out-of-the-box the only near universal attribute is probably `uncredited`, though even that should be amenable to administrator customization. Forethought is often imperfect, and a stash-box may want to add or modify attributes in the future. It allows a stash-box the ability to conform to its unique use-case.

These attributes should be implemented in such a way that each attribute corresponds to an ID. Therefore making modification trivial if the administrators decide they want to change the name of an attribute, e.g., `camo` to `cameo`, it will not require modification to each scene.
